src/machine/machine.c

```diff
@@ -57,27 +54,42 @@ int machine_new(Manager *manager, MachineClass class, const char *name, Machine
 
         *m = (Machine) {
                 .leader = PIDREF_NULL,
+                .vsock_cid = VMADDR_CID_ANY,
         };
 
-        m->name = strdup(name);
-        if (!m->name)
-                return -ENOMEM;
-
-        if (class != MACHINE_HOST) {
-                m->state_file = path_join("/run/systemd/machines", m->name);
-                if (!m->state_file)
+        if (name) {
+                m->name = strdup(name);
+                if (!m->name)
                         return -ENOMEM;
         }
 
         m->class = class;
 
-        r = hashmap_put(manager->machines, m->name, m);
+        *ret = TAKE_PTR(m);
+        return 0;
+}
+
+int machine_link(Manager *manager, Machine *machine) {
+        int r;
+        assert(manager);
+        assert(machine);
+        assert_return(!machine->manager, -EEXIST);
+        assert_return(machine->name, -EINVAL);
```

Comment:
assert_return is for libsystemd functions, just normal check here or normal assert